File: frontend/android/app/src/main/java/com/colebeing/ping/DeviceTokenStore.kt
package com.colebeing.ping

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

/**
 * Holds the long-lived device token the backend mints when the app registers its FCM token
 * (see /api/push/register-fcm). Encrypted at rest since it authenticates as the full user
 * identity with no expiry — read by NotificationActionReceiver, written by PingAuthPlugin.
 */
object DeviceTokenStore {
    private const val FILE_NAME = "ping_device_token"
    private const val KEY = "device_token"

    private fun prefs(context: Context): SharedPreferences {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        return EncryptedSharedPreferences.create(
            context,
            FILE_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    fun store(context: Context, token: String) {
        prefs(context).edit().putString(KEY, token).apply()
    }

    fun read(context: Context): String? {
        return prefs(context).getString(KEY, null)
    }
}

File: frontend/ios/App/App/DeviceTokenStore.swift
import Foundation
import Security

/// Holds the long-lived device token the backend mints when the app registers its FCM token
/// (see /api/push/register-fcm). Keychain-backed since it authenticates as the full user identity
/// with no expiry — the iOS analog of Android's EncryptedSharedPreferences-backed DeviceTokenStore.kt.
/// Read by PingNotificationDelegate, written by PingAuthPlugin.
enum DeviceTokenStore {
    private static let service = "com.colebeing.ping.devicetoken"
    private static let account = "device_token"

    static func store(_ token: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)

        var attributes = query
        attributes[kSecValueData as String] = Data(token.utf8)
        SecItemAdd(attributes as CFDictionary, nil)
    }

    static func read() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
              let data = result as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }
}
